File: WinForms/C#/Viewer/SearchForm.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;
using TatukGIS.NDK;

namespace Viewer
{
    /// <summary>
    /// Summary description for WinForm1.
    /// </summary>
    public class SearchForm : System.Windows.Forms.Form
    {
        /// <summary>
        /// Required designer variable.
        /// </summary>
        private System.ComponentModel.Container components = null;
        private System.Windows.Forms.StatusStrip stsBar;
        private System.Windows.Forms.ToolStripLabel toolStripLabel1;
        private System.Windows.Forms.ToolStripLabel toolStripLabel2;
        private System.Windows.Forms.Label label1;
        private System.Windows.Forms.Label label2;
        private System.Windows.Forms.Button btnSearch;
        private System.Windows.Forms.Label label3;
        private System.Windows.Forms.Label label4;
        private System.Windows.Forms.ComboBox cbLayer;
        private System.Windows.Forms.ComboBox cbField;
        private System.Windows.Forms.TextBox eValue;
        private System.Windows.Forms.ComboBox cbOperation;
        private System.Windows.Forms.GroupBox rgExtent;
        private System.Windows.Forms.RadioButton rbVisibleExtent;
        private System.Windows.Forms.RadioButton rbFullExtent;
        public WinForm mainForm;

        public SearchForm()
        {
            //
            // Required for Windows Form Designer support
            //
            InitializeComponent();

            //
            // TODO: Add any constructor code after InitializeComponent call
            //
            mainForm = null;
        }

        /// <summary>
        /// Clean up any resources being used.
        /// </summary>
        protected override void Dispose(bool disposing)
        {
            if (disposing)
            {
                if (components != null)
                {
                    components.Dispose();
                }
            }
            base.Dispose(disposing);
        }

        #region Windows Form Designer generated code
        /// <summary>
        /// Required method for Designer support - do not modify
        /// the contents of this method with the code editor.
        /// </summary>
        private void InitializeComponent()
        {
            this.stsBar = new System.Windows.Forms.StatusStrip();
            this.toolStripLabel1 = new System.Windows.Forms.ToolStripLabel();
            this.toolStripLabel2 = new System.Windows.Forms.ToolStripLabel();
            this.label1 = new System.Windows.Forms.Label();
            this.label2 = new System.Windows.Forms.Label();
            this.btnSearch = new System.Windows.Forms.Button();
            this.label3 = new System.Windows.Forms.Label();
            this.label4 = new System.Windows.Forms.Label();
            this.cbLayer = new System.Windows.Forms.ComboBox();
            this.cbField = new System.Windows.Forms.ComboBox();
            this.eValue = new System.Windows.Forms.TextBox();
            this.cbOperation = new System.Windows.Forms.ComboBox();
            this.rgExtent = new System.Windows.Forms.GroupBox();
            this.rbFullExtent = new System.Windows.Forms.RadioButton();
            this.rbVisibleExtent = new System.Windows.Forms.RadioButton();
            this.rgExtent.SuspendLayout();
            this.SuspendLayout();
            // 
            // stsBar
            // 
            this.stsBar.Location = new System.Drawing.Point(0, 77);
            this.stsBar.Name = "stsBar";
            this.stsBar.Items.AddRange(new System.Windows.Forms.ToolStripLabel[] {
            this.toolStripLabel1,
            this.toolStripLabel2});
            this.stsBar.Size = new System.Drawing.Size(493, 19);
            this.stsBar.TabIndex = 0;
            // 
            // toolStripLabel1
            // 
            this.toolStripLabel1.Name = "toolStripLabel1";
            this.toolStripLabel1.Text = " Layer :";
            this.toolStripLabel1.Width = 50;
            // 
            // toolStripLabel2
            // 
            this.toolStripLabel2.Name = "toolStripLabel2";
            this.toolStripLabel2.Width = 426;
            // 
            // label1
            // 
            this.label1.Location = new System.Drawing.Point(2, 0);
            this.label1.Name = "label1";
            this.label1.Size = new System.Drawing.Size(46, 13);
            this.label1.TabIndex = 1;
            this.label1.Text = "Layers : ";
            // 
            // label2
            // 
            this.label2.Location = new System.Drawing.Point(2, 40);
            this.label2.Name = "label2";
            this.label2.Size = new System.Drawing.Size(46, 13);
            this.label2.TabIndex = 2;
            this.label2.Text = "Fields : ";
            // 
            // btnSearch
            // 
            this.btnSearch.Cursor = System.Windows.Forms.Cursors.Hand;
            this.btnSearch.Font = new System.Drawing.Font("Verdana", 8.25F, System.Drawing.FontStyle.Bold, System.Drawing.GraphicsUnit.Point, ((byte)(238)));
            this.btnSearch.Location = new System.Drawing.Point(386, 46);
            this.btnSearch.Name = "btnSearch";
            this.btnSearch.Size = new System.Drawing.Size(103, 32);
            this.btnSearch.TabIndex = 3;
            this.btnSearch.Text = "Search";
            this.btnSearch.Click += new System.EventHandler(this.btnSearch_Click);
            // 
            // label3
            // 
            this.label3.Location = new System.Drawing.Point(226, 40);
            this.label3.Name = "label3";
            this.label3.Size = new System.Drawing.Size(46, 13);
            this.label3.TabIndex = 4;
            this.label3.Text = "Value : ";
            // 
            // label4
            // 
            this.label4.Location = new System.Drawing.Point(167, 40);
            this.label4.Name = "label4";
            this.label4.Size = new System.Drawing.Size(55, 13);
            this.label4.TabIndex = 5;
            this.label4.Text = "Operation : ";
            // 
            // cbLayer
            // 
            this.cbLayer.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList;
            this.cbLayer.Font = new System.Drawing.Font("Verdana", 8.25F, System.Drawing.FontStyle.Regular, System.Drawing.GraphicsUnit.Point, ((byte)(238)));
            this.cbLayer.ItemHeight = 13;
            this.cbLayer.Location = new System.Drawing.Point(2, 16);
            this.cbLayer.Name = "cbLayer";
            this.cbLayer.Size = new System.Drawing.Size(375, 21);
            this.cbLayer.TabIndex = 6;
            this.cbLayer.SelectedIndexChanged += new System.EventHandler(this.cbLayer_SelectedIndexChanged);
            // 
            // cbField
            // 
            this.cbField.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList;
            this.cbField.Font = new System.Drawing.Font("Verdana", 8.25F, System.Drawing.FontStyle.Regular, System.Drawing.GraphicsUnit.Point, ((byte)(238)));
            this.cbField.Location = new System.Drawing.Point(2, 56);
            this.cbField.Name = "cbField";
            this.cbField.Size = new System.Drawing.Size(161, 21);
            this.cbField.TabIndex = 7;
            // 
            // eValue
            // 
            this.eValue.Location = new System.Drawing.Point(224, 56);
            this.eValue.Name = "eValue";
            this.eValue.Size = new System.Drawing.Size(153, 20);
            this.eValue.TabIndex = 8;
            this.eValue.KeyPress += new System.Windows.Forms.KeyPressEventHandler(this.eValue_KeyPress);
            // 
            // cbOperation
            // 
            this.cbOperation.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList;
            this.cbOperation.Items.AddRange(new object[] {
            "=",
            "<>",
            ">",
            "<",
            ">=",
            "<="});
            this.cbOperation.Location = new System.Drawing.Point(168, 56);
            this.cbOperation.Name = "cbOperation";
            this.cbOperation.Size = new System.Drawing.Size(49, 21);
            this.cbOperation.TabIndex = 9;
            // 
            // rgExtent
            // 
            this.rgExtent.Controls.Add(this.rbFullExtent);
            this.rgExtent.Controls.Add(this.rbVisibleExtent);
            this.rgExtent.Location = new System.Drawing.Point(384, -2);
            this.rgExtent.Name = "rgExtent";
            this.rgExtent.Size = new System.Drawing.Size(105, 44);
            this.rgExtent.TabIndex = 10;
            this.rgExtent.TabStop = false;
            // 
            // rbFullExtent
            // 
            this.rbFullExtent.Location = new System.Drawing.Point(8, 25);
            this.rbFullExtent.Name = "rbFullExtent";
            this.rbFullExtent.Size = new System.Drawing.Size(92, 17);
            this.rbFullExtent.TabIndex = 1;
            this.rbFullExtent.Text = "Full Extent";
            this.rbFullExtent.CheckedChanged += new System.EventHandler(this.rbFullExtent_CheckedChanged);
            // 
            // rbVisibleExtent
            // 
            this.rbVisibleExtent.Location = new System.Drawing.Point(8, 8);
            this.rbVisibleExtent.Name = "rbVisibleExtent";
            this.rbVisibleExtent.Size = new System.Drawing.Size(92, 17);
            this.rbVisibleExtent.TabIndex = 0;
            this.rbVisibleExtent.Text = "Visible Extent";
            this.rbVisibleExtent.CheckedChanged += new System.EventHandler(this.rbVisibleExtent_CheckedChanged);
            // 
            // SearchForm
            // 
            this.AutoScaleDimensions = new System.Drawing.SizeF(96F, 96F);
            this.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Dpi;
            this.ClientSize = new System.Drawing.Size(493, 96);
            this.Controls.Add(this.rgExtent);
            this.Controls.Add(this.cbOperation);
            this.Controls.Add(this.eValue);
            this.Controls.Add(this.cbField);
            this.Controls.Add(this.cbLayer);
            this.Controls.Add(this.label4);
            this.Controls.Add(this.label3);
            this.Controls.Add(this.btnSearch);
            this.Controls.Add(this.label2);
            this.Controls.Add(this.label1);
            this.Controls.Add(this.stsBar);
            this.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedToolWindow;
            this.Location = new System.Drawing.Point(533, 173);
            this.Name = "SearchForm";
            this.StartPosition = System.Windows.Forms.FormStartPosition.CenterParent;
            this.Text = "Search";
            this.Load += new System.EventHandler(this.SearchForm_Load);
            this.Leave += new System.EventHandler(this.SearchForm_Leave);
            this.rgExtent.ResumeLayout(false);
            this.ResumeLayout(false);
            this.PerformLayout();

        }
        #endregion

        private void SearchForm_Load(object sender, System.EventArgs e)
        {
            int i, j;
            TGIS_LayerVector lv;

            if (mainForm.GIS.IsEmpty) return;

            cbLayer.Items.Clear();
            cbField.Items.Clear();
            btnSearch.Enabled = false;
            cbOperation.SelectedIndex = 0;
            rbVisibleExtent.Checked = true;

            // find all layers and make a list
            for (i = 0; i < mainForm.GIS.Items.Count; i++)
            {
                if (!(mainForm.GIS.Items[i] is TGIS_LayerVector))
                    continue;
                lv = (TGIS_LayerVector)mainForm.GIS.Items[i];
                if (lv.Name == null)
                    lv.Name = "[empty layer name]";
                cbLayer.Items.Add(lv.Name);
            }

            if (cbLayer.Items.Count > 0)
            {
                cbLayer.SelectedIndex = 0;

                //for first layer get fields names
                lv = (TGIS_LayerVector)mainForm.GIS.Items[cbLayer.SelectedIndex];
                for (j = 0; j < lv.Fields.Count; j++)
                {
                    if (lv.FieldInfo(j).Deleted) continue;
                    cbField.Items.Add(lv.FieldInfo(j).NewName);
                }

                cbField.SelectedIndex = 0;
                btnSearch.Enabled = true;
                stsBar.Items[1].Text = cbLayer.Items[cbLayer.SelectedIndex].ToString();
            }
            else
            {
                stsBar.Items[1].Text = "There is no Vector layer in the Viewer";
            }
        }

        private void SearchForm_Leave(object sender, System.EventArgs e)
        {
            mainForm.GIS.RevertAll();
            mainForm.GIS.FullExtent();
        }

        private void cbLayer_SelectedIndexChanged(object sender, System.EventArgs e)
        {
            int j;

            string name = (string)cbLayer.Items[cbLayer.SelectedIndex];

            // if layer changed, reload fields names
            // .Items[cbLayer.SelectedIndex]
            cbField.Items.Clear();
            for (j = 0;
                        j < ((TGIS_LayerVector)mainForm.GIS.Get(name)).Fields.Count;
                        j++)
            {
                if (((TGIS_LayerVector)mainForm.GIS.Get(name)).FieldInfo(j).Deleted) continue;
                cbField.Items.Add(((TGIS_LayerVector)mainForm.GIS.Get(name)).FieldInfo(j).NewName);
            }

            if (cbField.Items.Count > 0) cbField.SelectedIndex = 0;
            stsBar.Items[1].Text = cbLayer.Items[cbLayer.SelectedIndex].ToString();
        }

        private void eValue_KeyPress(object sender, System.Windows.Forms.KeyPressEventArgs e)
        {
            if (e.KeyChar == '\n')
                btnSearch_Click(this, e);
        }

        private void btnSearch_Click(object sender, System.EventArgs e)
        {
            TGIS_Shape shp;
            TGIS_LayerVector ll;
            TGIS_Extent ex;
            TGIS_FieldInfo fld;
            string sql;

            // get selected layer
            ll = (TGIS_LayerVector)mainForm.GIS.Items[cbLayer.SelectedIndex];

            // check if assigned, has proper type and searching value is not empty
            if ((ll == null) ||
                    ((TGIS_LayerAbstract)ll is TGIS_LayerPixel) ||
                    (eValue.Text == "")) return;

            // check the extent
            if (rbVisibleExtent.Checked)
                ex = mainForm.GIS.VisibleExtent;
            else
                ex = TGIS_Utils.GisWholeWorld();

            // calculate the condition similar to SQL where clause
            fld = ll.FieldInfo(ll.FindField(cbField.Items[cbField.SelectedIndex].ToString()));
            if (fld == null) return;
            if (fld.FieldType == TGIS_FieldType.String)
            {
                sql = cbField.Items[cbField.SelectedIndex].ToString() +
                      cbOperation.Items[cbOperation.SelectedIndex].ToString() +
                      "'" + eValue.Text + "'";
            }
            else
            {
                sql = cbField.Items[cbField.SelectedIndex].ToString() +
                      cbOperation.Items[cbOperation.SelectedIndex].ToString() +
                      eValue.Text;
            }

            // let's find any shapes meeting the criteria and flash them
            shp = ll.FindFirst(ex, sql);
            while (shp != null)
            {
                shp.Flash();
                Application.DoEvents();
                shp = ll.FindNext();
            }

            mainForm.GIS.Update();
        }

        private void rbVisibleExtent_CheckedChanged(object sender, System.EventArgs e)
        {
            rbFullExtent.Checked = !rbVisibleExtent.Checked;
        }

        private void rbFullExtent_CheckedChanged(object sender, System.EventArgs e)
        {
            rbVisibleExtent.Checked = !rbFullExtent.Checked;
        }
    }
}

File: WinForms/VB/Viewer/SearchForm.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Data
Imports TatukGIS.NDK

Namespace Viewer
    ''' <summary>
    ''' Summary description for WinForm1.
    ''' </summary>
    Public Class SearchForm
        Inherits System.Windows.Forms.Form
        ''' <summary>
        ''' Required designer variable.
        ''' </summary>
        Private components As System.ComponentModel.Container = Nothing
        Private stsBar As System.Windows.Forms.StatusStrip
        Private statusBarPanel1 As System.Windows.Forms.ToolStripStatusLabel
        Private statusBarPanel2 As System.Windows.Forms.ToolStripStatusLabel
        Private label1 As System.Windows.Forms.Label
        Private label2 As System.Windows.Forms.Label
        Private WithEvents btnSearch As System.Windows.Forms.Button
        Private label3 As System.Windows.Forms.Label
        Private label4 As System.Windows.Forms.Label
        Private WithEvents cbLayer As System.Windows.Forms.ComboBox
        Private cbField As System.Windows.Forms.ComboBox
        Private WithEvents eValue As System.Windows.Forms.TextBox
        Private cbOperation As System.Windows.Forms.ComboBox
        Private rgExtent As System.Windows.Forms.GroupBox
        Private WithEvents rbVisibleExtent As System.Windows.Forms.RadioButton
        Private WithEvents rbFullExtent As System.Windows.Forms.RadioButton
        Public mainForm As WinForm

        Public Sub New()
            '
            ' Required for Windows Form Designer support
            '
            InitializeComponent()

            '
            ' TODO: Add any constructor code after InitializeComponent call
            '
            mainForm = Nothing
        End Sub

        ''' <summary>
        ''' Clean up any resources being used.
        ''' </summary>
        Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
            If disposing Then
                If Not components Is Nothing Then
                    components.Dispose()
                End If
            End If
            MyBase.Dispose(disposing)
        End Sub

#Region "Windows Form Designer generated code"
        ''' <summary>
        ''' Required method for Designer support - do not modify
        ''' the contents of this method with the code editor.
        ''' </summary>
        Private Sub InitializeComponent()
            Me.stsBar = New System.Windows.Forms.StatusStrip()
            Me.statusBarPanel1 = New System.Windows.Forms.ToolStripStatusLabel()
            Me.statusBarPanel2 = New System.Windows.Forms.ToolStripStatusLabel()
            Me.label1 = New System.Windows.Forms.Label()
            Me.label2 = New System.Windows.Forms.Label()
            Me.btnSearch = New System.Windows.Forms.Button()
            Me.label3 = New System.Windows.Forms.Label()
            Me.label4 = New System.Windows.Forms.Label()
            Me.cbLayer = New System.Windows.Forms.ComboBox()
            Me.cbField = New System.Windows.Forms.ComboBox()
            Me.eValue = New System.Windows.Forms.TextBox()
            Me.cbOperation = New System.Windows.Forms.ComboBox()
            Me.rgExtent = New System.Windows.Forms.GroupBox()
            Me.rbFullExtent = New System.Windows.Forms.RadioButton()
            Me.rbVisibleExtent = New System.Windows.Forms.RadioButton()
            
            
            Me.rgExtent.SuspendLayout()
            Me.SuspendLayout()
            '
            'stsBar
            '
            Me.stsBar.Location = New System.Drawing.Point(0, 77)
            Me.stsBar.Name = "stsBar"
            Me.stsBar.Items.AddRange(New System.Windows.Forms.ToolStripStatusLabel() {Me.statusBarPanel1, Me.statusBarPanel2})

            Me.stsBar.Size = New System.Drawing.Size(493, 19)
            Me.stsBar.TabIndex = 0
            '
            'statusBarPanel1
            '
            
            Me.statusBarPanel1.Name = "statusBarPanel1"
            Me.statusBarPanel1.Text = " Layer :"
            Me.statusBarPanel1.Width = 50
            '
            'statusBarPanel2
            '
            Me.statusBarPanel2.AutoSize = True
            Me.statusBarPanel2.Name = "statusBarPanel2"
            Me.statusBarPanel2.Width = 426
            '
            'label1
            '
            Me.label1.Location = New System.Drawing.Point(2, 0)
            Me.label1.Name = "label1"
            Me.label1.Size = New System.Drawing.Size(46, 13)
            Me.label1.TabIndex = 1
            Me.label1.Text = "Layers : "
            '
            'label2
            '
            Me.label2.Location = New System.Drawing.Point(2, 40)
            Me.label2.Name = "label2"
            Me.label2.Size = New System.Drawing.Size(46, 13)
            Me.label2.TabIndex = 2
            Me.label2.Text = "Fields : "
            '
            'btnSearch
            '
            Me.btnSearch.Cursor = System.Windows.Forms.Cursors.Hand
            Me.btnSearch.Font = New System.Drawing.Font("Verdana", 8.25!, System.Drawing.FontStyle.Bold, System.Drawing.GraphicsUnit.Point, CType(238, Byte))
            Me.btnSearch.Location = New System.Drawing.Point(386, 46)
            Me.btnSearch.Name = "btnSearch"
            Me.btnSearch.Size = New System.Drawing.Size(103, 32)
            Me.btnSearch.TabIndex = 3
            Me.btnSearch.Text = "Search"
            '
            'label3
            '
            Me.label3.Location = New System.Drawing.Point(226, 40)
            Me.label3.Name = "label3"
            Me.label3.Size = New System.Drawing.Size(46, 13)
            Me.label3.TabIndex = 4
            Me.label3.Text = "Value : "
            '
            'label4
            '
            Me.label4.Location = New System.Drawing.Point(167, 40)
            Me.label4.Name = "label4"
            Me.label4.Size = New System.Drawing.Size(55, 13)
            Me.label4.TabIndex = 5
            Me.label4.Text = "Operation : "
            '
            'cbLayer
            '
            Me.cbLayer.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList
            Me.cbLayer.Font = New System.Drawing.Font("Verdana", 8.25!, System.Drawing.FontStyle.Regular, System.Drawing.GraphicsUnit.Point, CType(238, Byte))
            Me.cbLayer.ItemHeight = 13
            Me.cbLayer.Location = New System.Drawing.Point(2, 16)
            Me.cbLayer.Name = "cbLayer"
            Me.cbLayer.Size = New System.Drawing.Size(375, 21)
            Me.cbLayer.TabIndex = 6
            '
            'cbField
            '
            Me.cbField.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList
            Me.cbField.Font = New System.Drawing.Font("Verdana", 8.25!, System.Drawing.FontStyle.Regular, System.Drawing.GraphicsUnit.Point, CType(238, Byte))
            Me.cbField.Location = New System.Drawing.Point(2, 56)
            Me.cbField.Name = "cbField"
            Me.cbField.Size = New System.Drawing.Size(161, 21)
            Me.cbField.TabIndex = 7
            '
            'eValue
            '
            Me.eValue.Location = New System.Drawing.Point(224, 56)
            Me.eValue.Name = "eValue"
            Me.eValue.Size = New System.Drawing.Size(153, 20)
            Me.eValue.TabIndex = 8
            '
            'cbOperation
            '
            Me.cbOperation.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList
            Me.cbOperation.Items.AddRange(New Object() {"=", "<>", ">", "<", ">=", "<="})
            Me.cbOperation.Location = New System.Drawing.Point(168, 56)
            Me.cbOperation.Name = "cbOperation"
            Me.cbOperation.Size = New System.Drawing.Size(49, 21)
            Me.cbOperation.TabIndex = 9
            '
            'rgExtent
            '
            Me.rgExtent.Controls.Add(Me.rbFullExtent)
            Me.rgExtent.Controls.Add(Me.rbVisibleExtent)
            Me.rgExtent.Location = New System.Drawing.Point(384, -2)
            Me.rgExtent.Name = "rgExtent"
            Me.rgExtent.Size = New System.Drawing.Size(105, 44)
            Me.rgExtent.TabIndex = 10
            Me.rgExtent.TabStop = False
            '
            'rbFullExtent
            '
            Me.rbFullExtent.Location = New System.Drawing.Point(8, 25)
            Me.rbFullExtent.Name = "rbFullExtent"
            Me.rbFullExtent.Size = New System.Drawing.Size(92, 17)
            Me.rbFullExtent.TabIndex = 1
            Me.rbFullExtent.Text = "Full Extent"
            '
            'rbVisibleExtent
            '
            Me.rbVisibleExtent.Location = New System.Drawing.Point(8, 8)
            Me.rbVisibleExtent.Name = "rbVisibleExtent"
            Me.rbVisibleExtent.Size = New System.Drawing.Size(92, 17)
            Me.rbVisibleExtent.TabIndex = 0
            Me.rbVisibleExtent.Text = "Visible Extent"
            '
            'SearchForm
            '
            Me.AutoScaleDimensions = New System.Drawing.SizeF(96.0!, 96.0!)
            Me.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Dpi
            Me.ClientSize = New System.Drawing.Size(493, 96)
            Me.Controls.Add(Me.rgExtent)
            Me.Controls.Add(Me.cbOperation)
            Me.Controls.Add(Me.eValue)
            Me.Controls.Add(Me.cbField)
            Me.Controls.Add(Me.cbLayer)
            Me.Controls.Add(Me.label4)
            Me.Controls.Add(Me.label3)
            Me.Controls.Add(Me.btnSearch)
            Me.Controls.Add(Me.label2)
            Me.Controls.Add(Me.label1)
            Me.Controls.Add(Me.stsBar)
            Me.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedToolWindow
            Me.Location = New System.Drawing.Point(533, 173)
            Me.Name = "SearchForm"
            Me.StartPosition = System.Windows.Forms.FormStartPosition.CenterParent
            Me.Text = "Search"
            
            
            Me.rgExtent.ResumeLayout(False)
            Me.ResumeLayout(False)
            Me.PerformLayout()

        End Sub
#End Region

        Private Sub SearchForm_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Load
            Dim i, j As Integer
            Dim lv As TGIS_LayerVector

            If mainForm.GIS.IsEmpty Then
                Return
            End If

            cbLayer.Items.Clear()
            cbField.Items.Clear()
            btnSearch.Enabled = False
            cbOperation.SelectedIndex = 0
            rbVisibleExtent.Checked = True

            ' find all layers and make a list
            i = 0
            Do While i < mainForm.GIS.Items.Count
                If Not (TypeOf mainForm.GIS.Items(i) Is TGIS_LayerVector) Then
                    i += 1
                    Continue Do
                End If

                lv = CType(mainForm.GIS.Items(i), TGIS_LayerVector)

                If lv.Name Is Nothing Then
                    lv.Name = "[empty layer name]"
                End If
                cbLayer.Items.Add(lv.Name)
                i += 1
            Loop

            If cbLayer.Items.Count > 0 Then
                cbLayer.SelectedIndex = 0

                'for first layer get fields names
                lv = CType(mainForm.GIS.Items(cbLayer.SelectedIndex), TGIS_LayerVector)
                j = 0
                Do While j < lv.Fields.Count
                    If lv.FieldInfo(j).Deleted Then
                        GoTo Continue1
                    End If
                    cbField.Items.Add(lv.FieldInfo(j).NewName)
Continue1:
                    j += 1
                Loop

                cbField.SelectedIndex = 0
                btnSearch.Enabled = True
                stsBar.Items(1).Text = cbLayer.Items(cbLayer.SelectedIndex).ToString()
            Else
                stsBar.Items(1).Text = "There is no Vector layer in the Viewer"
            End If
        End Sub

        Private Sub SearchForm_Leave(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Leave
            mainForm.GIS.RevertAll()
            mainForm.GIS.FullExtent()
        End Sub

        Private Sub cbLayer_SelectedIndexChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles cbLayer.SelectedIndexChanged
            Dim j As Integer
            Dim name As String

            ' if layer changed, reload fields names
            cbField.Items.Clear()
            j = 0
            name = cbLayer.Items(cbLayer.SelectedIndex)
            Do While j < (CType(mainForm.GIS.Get(name), TGIS_LayerVector)).Fields.Count
                If (CType(mainForm.GIS.Get(name), TGIS_LayerVector)).FieldInfo(j).Deleted Then
                    GoTo Continue1
                End If
                cbField.Items.Add((CType(mainForm.GIS.Get(name), TGIS_LayerVector)).FieldInfo(j).NewName)
Continue1:
                j += 1
            Loop

            If cbField.Items.Count > 0 Then
                cbField.SelectedIndex = 0
            End If
            stsBar.Items(1).Text = cbLayer.Items(cbLayer.SelectedIndex).ToString()
        End Sub

        Private Sub eValue_KeyPress(ByVal sender As Object, ByVal e As System.Windows.Forms.KeyPressEventArgs) Handles eValue.KeyPress
            If e.KeyChar = ControlChars.Lf Then
                btnSearch_Click(Me, e)
            End If
        End Sub

        Private Sub btnSearch_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles btnSearch.Click
            Dim shp As TGIS_Shape
            Dim ll As TGIS_LayerVector
            Dim ex As TGIS_Extent
            Dim fld As TGIS_FieldInfo
            Dim sql As String

            ' get selected layer
            ll = CType(mainForm.GIS.Items(cbLayer.SelectedIndex), TGIS_LayerVector)

            ' check if assigned, has proper type and searching value is not empty
            If (ll Is Nothing) OrElse (TypeOf CType(ll, TGIS_LayerAbstract) Is TGIS_LayerPixel) OrElse (eValue.Text = "") Then
                Return
            End If

            ' check the extent
            If rbVisibleExtent.Checked Then
                ex = mainForm.GIS.VisibleExtent
            Else
                ex = TGIS_Utils.GisWholeWorld()
            End If

            ' calculate the condition similar to SQL where clause
            fld = ll.FieldInfo(ll.FindField(cbField.Items(cbField.SelectedIndex).ToString()))
            If fld Is Nothing Then
                Return
            End If
            If fld.FieldType = TGIS_FieldType.String Then
                sql = cbField.Items(cbField.SelectedIndex).ToString() & cbOperation.Items(cbOperation.SelectedIndex).ToString() & "'" & eValue.Text & "'"
            Else
                sql = cbField.Items(cbField.SelectedIndex).ToString() & cbOperation.Items(cbOperation.SelectedIndex).ToString() & eValue.Text
            End If

            ' let's find any shapes meeting the criteria and flash them
            shp = ll.FindFirst(ex, sql)
            Do While Not shp Is Nothing
                shp.Flash()
                Application.DoEvents()
                shp = ll.FindNext()
            Loop

            mainForm.GIS.Update()
        End Sub

        Private Sub rbVisibleExtent_CheckedChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles rbVisibleExtent.CheckedChanged
            rbFullExtent.Checked = Not rbVisibleExtent.Checked
        End Sub

        Private Sub rbFullExtent_CheckedChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles rbFullExtent.CheckedChanged
            rbVisibleExtent.Checked = Not rbFullExtent.Checked
        End Sub
    End Class
End Namespace
